csds-client/client/util.go

```diff
@@ -0,0 +1,201 @@
+		var data map[string]interface{}
+		err = json.Unmarshal(js, &data)
+		if err != nil {
+			return fmt.Errorf("%v", err)
+		}
+
+		// parse each json object to proto
+		i := 0
+		for _, n := range data["node_matchers"].([]interface{}) {
+			x := &envoy_type_matcher.NodeMatcher{}
+
+			jsonString, err := json.Marshal(n)
+			if err != nil {
+				return fmt.Errorf("%v", err)
+			}
+			err = protojson.Unmarshal(jsonString, x)
+			if err != nil {
+				return fmt.Errorf("%v", err)
+			}
+
+			// merge the proto with existing proto from request_file
+			if i < len(*nms) {
+				proto.Merge((*nms)[i], x)
+			} else {
+				*nms = append(*nms, x)
+			}
+			i++
+		}
+	}
+	return nil
+}
+
+// parse gcp project number from metadata of nodematchers
+func ParseGCPProject(nms []*envoy_type_matcher.NodeMatcher) string {
+	for _, nm := range nms {
+		for _, mt := range nm.NodeMetadatas {
+			for _, path := range mt.Path {
+				if path.GetKey() == "TRAFFICDIRECTOR_GCP_PROJECT_NUMBER" {
+					return mt.Value.GetStringMatch().GetExact()
+				}
+			}
+		}
+	}
+	return ""
+}
+
+// implement protoregistry.ExtensionTypeResolver and protoregistry.MessageTypeResolver to resolve google.protobuf.Any types
+type TypeResolver struct{}
+
+func (r *TypeResolver) FindMessageByName(message protoreflect.FullName) (protoreflect.MessageType, error) {
+	return nil, protoregistry.NotFound
+}
+
+// link the message type url to the specific message type
+func (r *TypeResolver) FindMessageByURL(url string) (protoreflect.MessageType, error) {
+	if url == "type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager" {
+		httpConnectionManager := envoy_config_filter_network_http_connection_manager_v2.HttpConnectionManager{}
+		return httpConnectionManager.ProtoReflect().Type(), nil
+	}
+	if url == "type.googleapis.com/envoy.api.v2.Cluster" {
+		cluster := envoy_api_v2.Cluster{}
+		return cluster.ProtoReflect().Type(), nil
+	}
+	if url == "type.googleapis.com/envoy.api.v2.Listener" {
+		listener := envoy_api_v2.Listener{}
+		return listener.ProtoReflect().Type(), nil
+	}
+	if url == "type.googleapis.com/envoy.config.filter.http.router.v2.Router" {
+		router := envoy_config_filter_http_router_v2.Router{}
+		return router.ProtoReflect().Type(), nil
+	}
+	if url == "type.googleapis.com/envoy.api.v2.RouteConfiguration" {
+		routeConfiguration := envoy_api_v2.RouteConfiguration{}
+		return routeConfiguration.ProtoReflect().Type(), nil
+	}
+	return nil, protoregistry.NotFound
+}
+
+func (r *TypeResolver) FindExtensionByName(field protoreflect.FullName) (protoreflect.ExtensionType, error) {
+	return nil, protoregistry.NotFound
+}
+
+func (r *TypeResolver) FindExtensionByNumber(message protoreflect.FullName, field protoreflect.FieldNumber) (protoreflect.ExtensionType, error) {
+	return nil, protoregistry.NotFound
+}
+
+// post process response
+func ParseResponse(response *envoy_service_status_v2.ClientStatusResponse, fileName string) {
```

Comment:
Seems that this function not just processes, but prints out the response. Maybe its name should reflect it? I.e. PrintOutResponse?

Reply:
Renamed it to `printOutResponse()`